### PR TITLE
[New Model] Stable Audio 3 (medium) — text-to-audio diffusion (#3787)

### LINKED ISSUE #3787: [New Model]: stabilityai/stable-audio-3-medium (DiT text-to-audio for music and SFX)
### The model to consider.

[Stable Audio 3](https://github.com/Stability-AI/stable-audio-3), the next generation of Stable Audio from Stability AI:

- Medium (GPU/CUDA, primary serving target): https://huggingface.co/stabilityai/stable-audio-3-medium — 1.4B params, up to 380s
- Small-Music / Small-SFX (CPU, 433M each, up to 120s): https://huggingface.co/stabilityai/stable-audio-3-small-music, https://huggingface.co/stabilityai/stable-audio-3-small-sfx
- Full collection: https://huggingface.co/collections/stabilityai/stable-audio-3
- Reference implementation (MIT): https://github.com/Stability-AI/stable-audio-3
- Technical report: https://arxiv.org/abs/2605.17991

Stable Audio 3 Large (2.7B) is API-only and out of scope.

### The closest model vllm-omni already supports.

`stabilityai/stable-audio-open-1.0` (#2917). Stable Audio 3 is its direct successor: same diffusion-transformer family and the same text-to-audio task. The pipeline shape (text encoder -> DiT denoiser -> audio autoencoder decode) carries over, so the existing Stable Audio Open integration is the natural starting point.

### What's your difficulty of supporting the model you want?

A few things differ from `stable-audio-open-1.0`:

- **New autoencoder.** Stable Audio 3 ships SAME (Semantic-Acoustic Music Encoder): stereo, 44.1 kHz, 256-dim latents, with separate Small/Large variants. The decoder is new code and supports chunked decoding to cap VRAM; it is not the stable-audio-open VAE.
- **No diffusers pipeline.** Unlike `stable-audio-open-1.0` (which has `StableAudioPipeline` in diffusers), Stable Audio 3 only ships the standalone `stable_audio_3` package in the Stability repo. The DiT and autoencoder adapters must be ported from that repo rather than reused from diffusers.
- **Variable-length generation.** The model sizes latents to the requested duration instead of a fixed window, which affects how the scheduler batches requests.
- **Three inference modes.** Beyond text-to-audio it supports audio-to-audio editing and inpainting/continuation, both of which take a reference-audio input. Initial support can scope to text-to-audio only.
- **Misc.** Medium requires Flash Attention 2; LoRA adapters are stackable and runtime-adjustable, which may want a serving-side knob later.

### Use case and motivation

Stable Audio 3 is the current open-weight SOTA for text-to-audio music and sound-effect generation, and is a straight upgrade over `stable-audio-open-1.0` in quality, max length (up to 380s for Medium), and speed. It fits vllm-omni's existing diffusion text-to-audio path, and supporting it keeps audio-generation coverage current. The Medium checkpoint is modest (1.4B, ~6.5 GB peak VRAM), so it serves on a single commodity GPU.

### PR DESCRIPTION
**Pull request by wuli666**
## Summary

Adds **Stable Audio 3 (medium)** — `stabilityai/stable-audio-3-medium`, a 1.4B-parameter
DiT text-to-audio diffusion model (stereo 44.1 kHz, up to 380 s) — to vLLM-Omni.

Closes #3787.

The model is ported from the standalone [Stability-AI/stable-audio-3](https://github.com/Stability-AI/stable-audio-3)
package (MIT); it has no diffusers pipeline, so this follows the custom-model
(Path B) integration pattern. Initial scope is **text-to-audio**; audio-to-audio
editing and inpainting are deferred.

## Architecture

Four-layer port mirroring upstream, under `vllm_omni/diffusion/models/stable_audio_3/`:

```
StableAudio3Pipeline.forward (text -> audio)
  ├─ MultiConditioner          T5GemmaConditioner (t5gemma-b-b-ul2) + NumberConditioner
  ├─ ConditionedDiffusionModelWrapper   routes conditioning into the DiT
  │    └─ DiffusionTransformer  1.4B DiT (denoise loop, bf16)
  └─ AudioAutoencoder (SAME)    encoder/decoder + bottleneck + patched pretransform (fp32 decode)
```

- **Weights**: the checkpoint is upstream's `ConditionedDiffusionModelWrapper` state_dict;
  `load_weights` remaps the leading `model.` → `diffusion_model.` and copies directly
  (T5Gemma loads itself in `__init__`). Verified 997/997 tensors, zero shape mismatch.
- **Sampling**: `dpmpp-3m-sde` with a `LogSNRShift` schedule (upstream default).
- **Discovery**: `download_stable_audio_3.py` writes the `model_index.json` /
  `transformer/config.json` the engine needs (the HF repo ships only `model_config.json`).

## What works (verified on `stable-audio-3-medium`, RTX 4090)

- ✅ Offline text-to-audio (`examples/offline_inference/text_to_audio/text_to_audio.py`).
- ✅ Online serving via `/v1/audio/generate` (`--model-class-name StableAudio3Pipeline`).
- ✅ Variable-length latents (5 s → 60 s scales correctly).
- ✅ **HSDP** weight sharding (`--use-hsdp`) — matches stable-audio-open's parallelism level.
- ✅ CPU / layerwise offload (`SupportsComponentDiscovery`).
- ✅ Runs without `flash-attn` (falls back to torch SDPA / flex-attention).

## Test plan

Offline driver (point `STABLE_AUDIO_3_TEST_MODEL` / `--model` at a directory prepared by
`examples/offline_inference/stable_audio_3/download_stable_audio_3.py`):

```bash
python examples/offline_inference/text_to_audio/text_to_audio.py \
    --model ./stable-audio-3-medium \
    --prompt "An ambient drone with shimmering overtones" \
    --audio-length 30.0 --num-inference-steps 100 --output sa3.wav
```

E2E (offline + online):

```bash
STABLE_AUDIO_3_TEST_MODEL=./stable-audio-3-medium pytest -v \
    tests/e2e/offline_inference/test_stable_audio_3_expansion.py \
    tests/e2e/online_serving/test_stable_audio_3_expansion.py
```

## Test result

E2E on a single RTX 4090: **offline 2 passed, online 1 passed**. Output is valid stereo
44.1 kHz audio (no NaN / clipping); the spectrum matches the prompt (sample below).

### Perf benchmark

**Model:** `stable-audio-3-medium` · **Steps:** 100 · **CFG:** 7.0 · **Sampler:** `dpmpp-3m-sde` · batch 1, warm (`torch.compile`), no `flash-attn` (PyTorch SDPA / flex-attention). RTF = E2E latency ÷ audio duration (lower is better).

| Config | Hardware | E2E Latency | RTF | Peak VRAM |
|--------|----------|-------------|-----|-----------|
| 30 s audio  | 1× RTX 4090 | 2.51 s | 0.084 | 5.7 GB |
| 60 s audio  | 1× RTX 4090 | 4.17 s | 0.070 | 5.7 GB |
| 120 s audio | 1× RTX 4090 | 7.90 s | 0.066 | 5.7 GB |

HSDP weight sharding is supported (`--use-hsdp`) but not benchmarked here (single-GPU box).

**Sample output** — 30 s, 100 steps, prompt *"warm analog synth pad with shimmering overtones"*:

[sa3_sample.wav](https://github.com/user-attachments/files/28377741/sa3_sample.wav)

## Not in scope (future work)

- TP / sequence-parallel / CFG-parallel — intentionally skipped; audio-diffusion
  peers (stable-audio-open, etc.) ship without them.
- Cache-DiT acceleration.
- audio-to-audio editing / inpainting (slots reserved, no user-facing entry yet).

## Hardware

`stable-audio-3-medium` runs on a single commodity GPU (~5.7 GB peak, measured above).
Medium prefers Flash Attention 2 but is not gated on it (SDPA fallback works).

<details>
<summary>Essential Elements of an Effective PR Description Checklist</summary>

- [x] The purpose of the PR + linked issue (Closes #3787).
- [x] The test plan (commands above).
- [x] The test results (e2e pass + generation-speed table).
- [x] (If applicable) Release notes / docs updated (example README + flow diagram).
</details>

### CODE REVIEW

**Comment by linyueqian** [comment]
please post perf benchmark as well

**Comment by wuli666** [comment]
> please post perf benchmark as well

@linyueqian PTAL

**Comment by wuli666** [comment]
@linyueqian Any updates on this PR?

**Review by linyueqian — COMMENTED** [review]
**E2E verification on H200**

Ran this PR's merge commit (head `08e898b1` merged into main `bf03413c`) on a single H200 with vLLM 0.22.0 and transformers 5.8.1, weights from the gated HF repo prepared via `download_stable_audio_3.py`'s layout.

Results:
- `tests/e2e/offline_inference/test_stable_audio_3_expansion.py`: 2 passed (108s), including the variable-length 5s vs 60s latent check
- `tests/e2e/online_serving/test_stable_audio_3_expansion.py`: 1 passed (36s) via `/v1/audio/generate`
- Quality spot check: 10s clip at 100 steps, CFG 7.0, seed 42. Valid stereo 44.1 kHz, no NaN, sounds good by ear.

Observations, none blocking:

1. [Minor] Since the model is gated, the e2e tests need an HF token that has accepted the Stability license. The test file imports `skip_if_gated_repo_inaccessible` (`noqa: F401`) but never calls it, so on a machine or CI runner without gate access the tests error at engine init instead of skipping cleanly. Suggest calling it when `STABLE_AUDIO_3_TEST_MODEL` is the HF repo id rather than a local directory.

2. [Minor] Docs surfaces: please add a `StableAudio3Pipeline` row to `docs/models/supported_models.md` and mention SA3 in `docs/user_guide/examples/offline_inference/text_to_audio.md` and `docs/user_guide/examples/online_serving/text_to_audio.md`, including the gated-access step the stable-audio-open page already documents. Since SA3 is the direct successor to stable-audio-open (#3787), the docs should position it as the primary text-to-audio model.

3. [Minor] The 100-step sample hard-clips on ~0.5% of samples (peak pinned at full scale). Looks like hot float output at CFG 7.0 getting clipped during int16 conversion rather than a model bug. Worth considering peak normalization before the int16 write, possibly as a follow-up since `text_to_audio.py` is shared.

4. [Nit] The README says medium "requires Flash Attention 2 at runtime", but my env had no `flash_attn` and the SDPA fallback worked end to end. Suggest softening to "recommended for performance".

Looking ahead: my expectation is that we deprecate `stable-audio-open-1.0` in favor of SA3 after this lands (separate issue to follow, since several in-flight PRs still target stable-audio-open). With that in mind, these e2e tests should take over the CI slot that was planned for the stable-audio-open suite.

Nice work on the port. The PR's verification claims reproduce on Hopper-class hardware against the current vLLM 0.22 stack, which the branch predates.

**Comment by linyueqian** [comment]
also need to add recipes

**Comment by linyueqian** [comment]
@hsliuustc0106 this is a newer version of stable audio. please take a look and see if you can add the model in CI's hugging face account. it requires consent.

File: examples/offline_inference/stable_audio_3/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+## Hardware
+
+| Variant | Peak VRAM | Notes |
+|---------|-----------|-------|
+| `stable-audio-3-medium` | ~6.5 GB | Single commodity GPU (RTX 3090 / 4070 / A100) |
+| `stable-audio-3-small-music` | CPU-targeted | Quality lower; use only when no GPU |
+| `stable-audio-3-small-sfx` | CPU-targeted | Quality lower; use only when no GPU |
+
+Medium requires **Flash Attention 2** at runtime.
```
Comment (linyueqian) [inline examples/offline_inference/stable_audio_3/README.md]: [Minor] The new recipe (`recipes/StabilityAI/Stable-Audio-3-Medium.md`) says Flash Attention 2 is "recommended for performance but not required", which matches what I observed during verification (the SDPA fallback worked end to end without `flash_attn` installed). This line still says Medium requires it at runtime, so the two docs now contradict each other. Please soften this one to match the recipe wording.

**Comment by hsliuustc0106** [comment]
> @hsliuustc0106 this is a newer version of stable audio. please take a look and see if you can add the model in CI's hugging face account. it requires consent.

@yenuo26 @congw729  @ywang96 PTAL,

**Comment by hsliuustc0106** [comment]
do we have a plan to replace the previous version with this model?

**Comment by wuli666** [comment]
@linyueqian @hsliuustc0106  anything update？
